### PR TITLE
Provide encoding/decoding instrumentation entrypoints

### PR DESCRIPTION
**Pull request by maximebedard**
I managed to be able to hook myself and do this properly by override #initialize on the memcached client, but this made it dependant on load order. Which is insanely problematic in shopify core.

Instead of depending on load order, I'm providing specific entry points instead.

### CODE REVIEW

**Review by shivnagarajan — APPROVED** [review]
seems reasonable to me. Is there a way to test/validate that does what we want, and not have any other side-effects?

**Comment by maximebedard** [comment]
I don't think so, but extracting that call into instance methods seems pretty safe to me. If there's any side-effect of doing this I'm not really sure what we can trust :)

**Review by tylermercier — APPROVED** [review]

**Review by inf-rno — APPROVED** [review]

File: lib/memcached/memcached.rb [hunk 1]
```diff
@@ -582,6 +582,16 @@ def stats(subcommand = nil)
...
 
+  # write instrumentation entrypoints
+  def do_encode(key, value, flags)
+    @codec.encode(key, value, flags)
+  end
+
+  # read instrumentation entrypoints
+  def do_decode(key, value, flags)
+    @codec.decode(key, value, flags)
+  end
```
Comment (dylanahsmith) [inline lib/memcached/memcached.rb:593]: codec is a library user provided value.  Why don't we instrument its `encode` and `decode` methods?  Then I don't think we would have any need for these methods that seem like they are for monkey patching.

Comment (maximebedard) [reply lib/memcached/memcached.rb:593]: I think the original intent was to be able to override theses methods in instrumentation gems (such as tracing or shopify_metrics) without having to override `#initialize`.

I was initially able to have something similar to this, but prepending that module became load order dependent, which was much more tricky to figure out and unpredictable.

```rb
module Patch
  def initialize(*)
    super
    @codec = InstrumentedCodec.new(@codec)
  end
end

Memcached.prepend(Patch) # <= load order dependent.
```

If you have any alternative solution, I'm all ears. But this seemed to be the less fragile way of doing it IMO.

Comment (dylanahsmith) [reply lib/memcached/memcached.rb:593]: The problem is that this is dependant on our fork being used.  If the upstream memcached gem is used, then the method overrides won't override anything or be called.  If we need to patch something, we should try to make our fork temporary by trying to upstream all our patches.

Comment (dylanahsmith) [reply lib/memcached/memcached.rb:593]: I don't think we need to make the instrumentation automatic.  For instance, it looks like we are now passing a `instrumentation_tags` option to `memcached_store`, which is actually more of an option to the instrumentation monkey patch.  We could more cleanly just pass in an InstrumentedCodec instance from the application and the `instrumentation_tags` option could instead be passed into `InstrumentedCodec.new`.

By the way, I have been added as a contributor to the memcached gem, so there shouldn't be any blockers to getting a change upstream if it makes sense there.  So feel free to open PRs there.

**Pull request merged**